nbextensions/bulkImportCell/bulkImportCell.js

```diff
@@ -189,34 +223,58 @@ define([
             // widgets this cell owns
             cellTabs,
             controlPanel,
-            fileTypePanel,
-            model = Props.make({
-                data: TestAppObj,
-                onUpdate: function (props) {
-                    Utils.setMeta(this.cell, 'bulkImportCell', props.getRawObject());
-                },
-            });
+            fileTypePanel;
         if (options.initialize) {
-            initialize(typesToFiles);
+            initialize(options.specs);
         }
+        let model = Props.make({
+            data: Utils.getMeta(cell, 'bulkImportCell'),
+            onUpdate: function(props) {
+                Utils.setMeta(cell, 'bulkImportCell', props.getRawObject());
+            }
+        });
 
-        let spec = Spec.make({
-            appSpec: model.getItem('app.spec'),
+        let specs = {};
+        let rawSpecs = model.getItem('app.specs');
+        Object.keys(rawSpecs).forEach((appId) => {
+            specs[appId] = Spec.make({
+                appSpec: rawSpecs[appId]
+            });
         });
 
         setupCell();
 
         /**
          * Does the initial pass on newly created cells to initialize its metadata and get it
          * set up for a new life as a Bulk Import Cell.
-         *
-         * @param {object} typesToFiles keys = data type strings, values = arrays of file paths
-         * to import
-         * e.g.: {
-         *     'fastq_reads': ['file1.fq', 'file2.fq']
-         * }
+         * @param {object} appSpecs - a mapping from app id -> app specs as defined by the
+         *  Narrative Method Store service
          */
-        function initialize(_typesToFiles) {
+        function initialize(appSpecs) {
+            /* Initialize the parameters section.
+             * This is broken down per-app, and here we use the file type
+             * as the key.
+             * So we need to filter through the parameters section of each spec,
+             * generate the parameter ids, and make a structure like:
+             * {
+             *   fileType1: {
+             *      param1: '' or default from spec,
+             *      param2: '' or default from spec,
+             *      ...etc.
+             *   },
+             *   fileType2: {...}
+             * }
+             *
+             */
+            const initialParams = {};
```

Comment:
This works for now, but will need some updating.
Right now, it treats each app as an individual run. When, really, the "inputs" should be done as lists, as represented in the filePathWidget. But each app has those input parameters as singletons. I'm not sure the right way to do this yet.

Reply:
don't have a good answer as this seems like a tough problem. is it worth representing here in a comment as a todo and/or a Jira ticket? 

how does batch mode handle it? is the only way to have a single app run process multiple files with a set of params for those files to update the app itself?